File: binarysearchtree.py
class Node:
    def __init__(self, data=None):
        self.data = data
        self.left = None
        self.right = None

    def __str__(self):
        return str(self.data)

    def __iter__(self):
        return self.inorder(self)

    def inorder(self, node):
        if node is None or node.data is None:
            return
        if node.left is not None:
            for x in self.inorder(node.left):
                yield x
        yield node
        if self.right is not None:
            for x in self.inorder(node.right):
                yield x
        pass


class BinarySearchTree:

    def __init__(self, name, root):
        self.name = name
        self.root = root

    def insert(self, root, newdata):

        if root is None or root.data is None:
            root = Node(newdata)
            return root
        elif root.data > newdata:
            root.left = self.insert(root.left, newdata)
        else:
            root.right = self.insert(root.right, newdata)
        return root

    def inserting(self, newdata):
        self.root = self.insert(self.root, newdata)

    def add_all(self, *list):
        for i in list:
            self.inserting(i)

    def __iter__(self):
        return self.inorder(self.root)

    def inorder(self, node):
        if node is None:
            return
        if node.left is not None:
            for x in self.inorder(node.left):
                yield x
        yield node
        if node.right is not None:
            for x in self.inorder(node.right):
                yield x

    def __str__(self):
        return "[" + self.name + "] " + self.root.__str__() + (self.printing(self.root)).__str__()

    def printing(self, current):
        x = ""
        if current.left is not None:
            x = x + " L:(" + current.left.__str__() + self.printing(current.left) + ")"
        if current.right is not None:
            x = x + " R:(" + current.right.__str__() + self.printing(current.right) + ")"
        return x

File: BinarySearchTree.java
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;

public class BinarySearchTree<T extends Comparable<T>> implements Iterable<BinarySearchTree.Node<T>> {

    public static class Node<T> {
        public T data;
        public Node<T> left;
        public Node<T> right;
        public Node<T> parent;

        public Node(T data) {
            this.data = data;
            this.left = null;
            this.right = null;
            this.parent = null;

        }
        @Override
        public String toString() {
            return data.toString();
        }
    }
    public Node<T> root;
    public String name;

    public BinarySearchTree(String name) {
        this.root = null;
        this.name = name;
    }

    public void insert(T newData) {
        this.root = insert(root, newData);
    }

    public Node<T> insert(Node<T> root, T newData) {
        if (root == null) {
            root = new Node<T>(newData);
            return root;
        }
        else if (root.data.compareTo(newData) >= 0) {
            root.left = insert(root.left, newData);
            root.left.parent = root;
        } else {
            root.right = insert(root.right, newData);
            root.right.parent = root;
        }
        return root;
    }
    public Node<T> getRoot(){
        return root;
    }
    @Override
    public Iterator<Node<T>> iterator() {
        return new myIterator<>(this);
    }

    @Override
    public String toString() {
        return "[" + this.name +"]" + "  " + root.toString() + printing(root);
    }

    private String printing(Node<T> current) {
        String x = "";
        if(current.left != null) {
            x = x + " L:(" + current.left.toString() + printing(current.left) + ")";
        }
        if(current.right != null) {
            x = x + " R:(" + current.right.toString() + printing(current.right) + ")";
        }
        return x;
    }

    public static class myIterator<T extends Comparable<T>> implements Iterator<Node<T>>{
        Node<T> current;
        public myIterator(BinarySearchTree<T> list)
        {
            current = list.getRoot();
            while (current.left != null)
                current = current.left;
        }
        @Override
        public boolean hasNext() {
            return current != null;
        }

        @Override
        public Node<T> next() {
            if(!hasNext()) throw new NoSuchElementException();
            Node<T> r = current;
            if(current.right != null) {
                current = current.right;
                while (current.left != null)
                    current = current.left;
                return r;
            }
            while(true) {
                if(current.parent == null) {
                    current = null;
                    return r;
                }
                if(current.parent.left == current) {
                    current = current.parent;
                    return r;
                }
                current = current.parent;
            }
        }
    }
    public void addAll(List<T> asList) {
        asList.forEach(this::insert);
    }
    private ArrayList<T> getList(){
        ArrayList<T> tree = new ArrayList<T>();
        this.iterator().forEachRemaining(a -> tree.add(a.data));
        return tree;
    }
    public static <T extends Comparable<T>> List<T> merge(List<BinarySearchTree<T>> bstlist){
        final ArrayList<T>[] master = new ArrayList[]{new ArrayList<>()};
        List<Thread> threads = new ArrayList<>();
        for(int i = 0; i < bstlist.size(); i++) {
            int finalI = i;
            Thread t1 = new Thread(new Runnable() {
                @Override
                public void run() {
                    ArrayList<T> tree = bstlist.get(finalI).getList();
                    synchronized(master[0]) {
                        master[0] = merging(tree, master[0]);
                    }
                }
                private <T extends Comparable<T>> ArrayList<T> merging(ArrayList<T> tree, ArrayList<T> master) {
                    ArrayList<T> temp = new ArrayList<>();
                    int i = 0;
                    int j = 0;
                    while (i < master.size() && j < tree.size()) {
                        if (master.get(i).compareTo(tree.get(j)) < 0) {
                            temp.add(master.get(i));
                            i++;
                        } else {
                            temp.add(tree.get(j));
                            j++;
                        }
                    }
                        if (i == master.size()) {
                            while (j < tree.size()) {
                                temp.add(tree.get(j));
                                j++;
                            }
                        }
                        if (j == tree.size()) {
                            while (i < master.size()) {
                                temp.add(master.get(i));
                                i++;
                            }
                        }
                    return temp;
                }
            });
        threads.add(t1);
        }
        for (Thread thread : threads) {
            thread.start();
        }
        for (Thread thread : threads) {
            try {
                thread.join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        return master[0];
    }
    }

